### DISCUSSION TITLE
How to read from buffer without 'Buffer in not mapped error' (wasm version, winit loop)

### DISCUSSION THREAD

**Discussion by jaankaup** [gfx-rs/wgpu#2593] [Q&A] [answered]
Why does this ṕeace of code work when compiling to native version but not when compiling to wasm?

https://github.com/jaankaup/jaankaup-engine/blob/main/jaankaup_core/src/buffer.rs#L61
61    let data = buffer_slice.get_mapped_range().to_vec();

This web page crashes at the code (line 61) (DOMException: GPUBuffer.getMappedRange: Buffer is not mapped).
https://jaankaup.github.io/curves/

This is an example of the marching cubes algorithm that seems to work on the latest version of Firefox Nightly (NVidia/win10/ubuntu).
https://jaankaup.github.io/mc/

It works because I used  draw_indirect to get around the map_async and get_mapped_range problem. However, It would be nice to read from buffers to the host program when compiling to wasm. Not only once when calling async function but on every frame update.

**Comment by grovesNL** [comment, marked as answer]
`poll` no-ops on the web (see https://docs.rs/wgpu/latest/wgpu/struct.Device.html#method.poll), so you won't be able to wait/synchronously block on the buffer to be mapped like you can natively.

Instead you'll need to `await` `map_async` for this to work on the web.
